noir-projects/noir-protocol-circuits/bootstrap.sh

```diff
@@ -51,73 +51,154 @@ mkdir -p $key_dir
+  # Get commands for writing vk and converting to fields based on name
+  if echo "$name" | grep -qE "${ivc_regex}"; then
     local write_vk_cmd="write_vk_for_ivc"
     local vk_as_fields_cmd="vk_as_fields_mega_honk"
   elif echo "$name" | grep -qE "${rollup_honk_regex}"; then
-    local proto="ultra_rollup_honk"
     local write_vk_cmd="write_vk_ultra_rollup_honk -h 2"
     local vk_as_fields_cmd="vk_as_fields_ultra_rollup_honk"
   else
-    local proto="ultra_honk"
     local write_vk_cmd="write_vk_ultra_honk -h 1"
     local vk_as_fields_cmd="vk_as_fields_ultra_honk"
   fi
-  echo "$proto$"
 
-  # No vks needed for simulated circuits.
-  [[ "$name" == *"simulated"* ]] && return
+  local filename="$name.json"
+  local json_path="./target/$filename"
+  local key_path="$key_dir/$name.vk.data.json"
+  echo_stderr "Generating vk for function: $name..."
 
   # Change this to add verification_key to original json, like contracts does.
   # Will require changing TS code downstream.
-  bytecode_hash=$(jq -r '.bytecode' $json_path | sha256sum | tr -d ' -')
-  hash=$(hash_str "$BB_HASH-$bytecode_hash-$proto")
-  if ! cache_download vk-$hash.tar.gz 1>&2; then
-    local key_path="$key_dir/$name.vk.data.json"
-    echo_stderr "Generating vk for function: $name..."
-    SECONDS=0
-    local vk_cmd="jq -r '.bytecode' $json_path | base64 -d | gunzip | $BB $write_vk_cmd -b - -o - --recursive | xxd -p -c 0"
-    echo_stderr $vk_cmd
-    vk=$(dump_fail "$vk_cmd")
-    local vkf_cmd="echo '$vk' | xxd -r -p | $BB $vk_as_fields_cmd -k - -o -"
-    # echo_stderrr $vkf_cmd
-    vk_fields=$(dump_fail "$vkf_cmd")
-    jq -n --arg vk "$vk" --argjson vkf "$vk_fields" '{keyAsBytes: $vk, keyAsFields: $vkf}' > $key_path
-    echo_stderr "Key output at: $key_path (${SECONDS}s)"
-    cache_upload vk-$hash.tar.gz $key_path &> /dev/null
-  fi
+  local vk vk_fields
+  SECONDS=0
+  local vk_cmd="jq -r '.bytecode' $json_path | base64 -d | gunzip | $BB $write_vk_cmd -b - -o - --recursive | xxd -p -c 0"
+  echo_stderr $vk_cmd
+  vk=$(dump_fail "$vk_cmd")
+  local vkf_cmd="echo '$vk' | xxd -r -p | $BB $vk_as_fields_cmd -k - -o -"
+  # echo_stderrr $vkf_cmd
+  vk_fields=$(dump_fail "$vkf_cmd")
+  jq -n --arg vk "$vk" --argjson vkf "$vk_fields" '{keyAsBytes: $vk, keyAsFields: $vkf}' > $key_path
+  echo_stderr "Key output at: $key_path (${SECONDS}s)"
+  cache_upload vk-$hash.tar.gz $key_path &> /dev/null
+
+  echo $name
+}
+
+# Compiles a single circuit
+function compile {
+  set -eu
+  [ -f "package.json" ] && denoise "yarn && node ./scripts/generate_variants.js"
+  local dir=$1
+  generate_vk $(try_download_vk $(compile_circuit $(try_download_circuit $dir)))
 }
 
+# Runs the following steps for each circuit:
+# 1. try_download_circuit
+# 2. compile_circuit
+# 3. try_download_vk
+# 4. generate_vk
 function build {
   set +e
   set -u
 
   [ -f "package.json" ] && denoise "yarn && node ./scripts/generate_variants.js"
+  local parallel_cmd="parallel -t --line-buffer --halt now,fail=1"
```

Comment:
I had to remove `--tag` since it polluted stdout. With `--tag` set, the input to the second command (`compile_circuit`) included the tag. However, by removing the tag we now don't get any feedback on what circuit are we running.

Any idea on how to make `--tag` only apply to stderr? Or should I arrange this in a different way? An easy solution is to manually include the circuit `name` in every `echo_stderr` we do here, since there are not too many. But I'd like to know what's the proper way of handling this.